CRISPResso2/CRISPRessoShared.py

```diff
@@ -1661,3 +1673,462 @@ def zip_results(results_folder):
         )
     sb.call(cmd_to_zip, shell=True)
     return
+
+def safety_check(crispresso2_info, aln_stats, logger=None, min_total_reads=10000, alignedCutoff=.9, alternateAlignment=.3, minRatioOfModsInToOut=.01, modificationsAtEnds=.01, outsideWindowMaxSubRate=.002, maxRateOfSubs=0.3, guide_len=19, amplicon_len=50, ampliconToReadLen = 1.5):
+    """Check the results of analysis for potential issues and warns the user.
+
+    Parameters
+    ----------
+    crispresso2_info : dict
+        Dictionary of values describing the analysis
+    aln_stats : dict
+        Dictionary of alignment statistics and modification frequency and location
+    logger : logger
+        Logger from the calling function for logging messages
+    min_total_reads : int
+        Cutoff value for total reads aligned
+    alignedCutoff : float
+        Check value for the percentage of reads aligned vs not aligned
+    alternateAlignment : float
+        Percentage of variance from expected reads to trigger guardrail
+    minRatioOfModsInToOut : float
+        Float representing the acceptable ratio of modifications in the window to out
+    modificationsAtEnds : float
+        The ratio of reads with modifications at the 0 or -1 spot
+    outsideWindowMaxSubRate : float
+        Ratio of subs allowed outside of the window
+    maxRateOfSubs : float
+        Allowed rate of subs accross the entire read
+    guide_len : int
+        Minimum guide length
+    amplicon_len : int 
+        Minimum guide length
+    ampliconToReadLen : float
+        Comparison value between amplicons and reads
+    """
+    if logger is None:
+        # this line noise will get the name of the module from which this
+        # function was called, and thereby the correct logger
+        logger = logging.getLogger(getmodule(stack()[1][0]).__name__)
+    messageHandler = GuardRailMessageHandler(logger)
+    
+    totalReadsGuardRail = TotalReadsGuardRail(messageHandler, min_total_reads)
+    totalReadsGuardRail.safety(aln_stats['N_TOT_READS'])
```

Comment:
You can totally disregard this comment, but what would you think about something like this?

```
guard_rails = [
    (TotalReadsGuardRail(messageHandler, min_total_reads), {'total': aln_stats['N_TOT_READS']}),
    (OverallReadsAlignedGuardRail(messageHandler, alignedCutoff), {'total_reads': aln_stats['N_TOT_READS'], 'n_reads_aligned': (aln_stats['N_CACHED_ALN'] + aln_stats['N_COMPUTED_ALN'])}),
    ...
]

map(lambda guard_rail, args: guard_rail.safety(**args), guard_rails)
```

Reply:
I've been trying to get this to work but for some reason the message handler doesn't function as expected and doesn't display the messages.

Reply:
Sorry about that, if it isn't working there is no need to fight it, we can stick with the original way that works.